### PR TITLE
sneak '--always-make' into R CMD SHLIB invocation

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

Addresses #6384.

### Approach

- Ensure `R CMD SHLIB` runs in the package `src/` directory directly.
- Sneak in the `--always-make` argument into the `R CMD SHLIB` invocation.

### Automated Tests

N/A

### QA Notes

N/A

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/session/modules/clang/RCompilationDatabase.cpp [hunk 1]
```diff
@@ -555,9 +565,62 @@ std::vector<std::string> RCompilationDatabase::compileArgsForPackage(
...
+      {
+         LOG_ERROR(error);
+         return {};
+      }
+   }
+   
+   // call R CMD shlib on that file
+   std::vector<std::string> compileArgs = argsForRCmdSHLIB(env, targetSrcFile);
+
+   // clena up
```
Comment (kevinushey) [inline src/cpp/session/modules/clang/RCompilationDatabase.cpp:619]:
```suggestion
   // clean up
```

File: src/cpp/session/modules/clang/RCompilationDatabase.cpp [hunk 2]
```diff
@@ -790,31 +853,45 @@ Error RCompilationDatabase::executeSourceCpp(
...
    
    // compile the file as dry-run
    module_context::RCommand rCmd(rBinDir);
    rCmd << "SHLIB";
    rCmd << "--dry-run";
-   rCmd << output;
+   rCmd << srcPath.getFilename();
+   
+   // a dirty trick to inject '--always-make' into the make invocation
+   rCmd << "\"' --always-make IGNORED='\"";
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/cpp/session/modules/clang/RCompilationDatabase.cpp:872]: Do we have a way to monitor that the hack would still work if `tools:::.shlib_internal()` was modified ?

**Review by romainfrancois — COMMENTED** [review]
(I don't really get the rtools/windows specific parts but it looks fine). 

I've tested this with `dplyr`, `vctrs` and `ps`  on macOS, and I do get auto-complete and the contents of `# PACKAGE COMPILATION ARGS ----` seems legit. 

Minor comment on the hack and the necessity to use a tempfile when no real file is found, but otherwise 🤞 it looks like this could be the end of this saga.

File: src/cpp/session/modules/clang/RCompilationDatabase.cpp [hunk 3]
```diff
@@ -555,9 +565,62 @@ std::vector<std::string> RCompilationDatabase::compileArgsForPackage(
...
+      {
+         targetSrcFile = srcFile;
+         break;
+      }
+   }
+
+   // if we couldn't find a source file, create and use a tempfile
+   // (presumedly this could happen in a brand new project?)
+   // (or should we be waiting for the user to create a file?)
+   bool removeOnExit = false;
```
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/cpp/session/modules/clang/RCompilationDatabase.cpp:602]: That's a good idea.

File: src/cpp/session/modules/clang/RCompilationDatabase.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/cpp/session/modules/clang/RCompilationDatabase.cpp:857]: I can't think of a great way... in theory we could use `digest()` and compare the hashes of the current version of the function with later versions, but it's also possible that the function might change in a way that's still compatible with this hack.

Another approach would be to use something like `trace()` to inject `--always-make` in the right place of `tools:::.shlib_internal()`, if we could figure out some heuristic for where that should be injected. But we can cross that bridge some day in the future if it ever is required.

**Pull request merged**